src/gui/DetailsWidget.cpp

```diff
@@ -32,325 +32,291 @@
+    m_ui->groupTabWidget->setCurrentIndex(tabIndex);
+}
 
-    auto icon = m_currentEntry->iconPixmap();
-    if (icon.width() > 16 || icon.height() > 16) {
-        icon = icon.scaled(16, 16);
+void DetailsWidget::setDatabaseMode(DatabaseWidget::Mode mode)
+{
+    m_locked = mode == DatabaseWidget::LockedMode;
+    if (m_locked) {
+        return;
     }
-    m_ui->entryIcon->setPixmap(icon);
-
-    QString title = QString(" / ");
-    Group* entry_group = m_currentEntry->group();
-    if (entry_group) {
-        QStringList hierarchy = entry_group->hierarchy();
-        hierarchy.removeFirst();
-        title += hierarchy.join(" / ");
-        if (hierarchy.size() > 0) {
-            title += " / ";
+
+    if (mode == DatabaseWidget::ViewMode) {
+        if (m_ui->stackedWidget->currentWidget() == m_ui->pageGroup) {
+            setGroup(m_currentGroup);
+        } else {
+            setEntry(m_currentEntry);
         }
     }
-    title.append(m_currentEntry->resolveMultiplePlaceholders(m_currentEntry->title()));
-    m_ui->titleLabel->setText(title);
+}
+
+void DetailsWidget::updateEntryHeaderLine()
+{
+    Q_ASSERT(m_currentEntry);
+    const QString title = m_currentEntry->resolveMultiplePlaceholders(m_currentEntry->title());
+    m_ui->entryTitleLabel->setText(hierarchy(m_currentEntry->group(), title));
+    m_ui->entryIcon->setPixmap(preparePixmap(m_currentEntry->iconPixmap(), 16));
+}
+
+void DetailsWidget::updateEntryTotp()
+{
+    Q_ASSERT(m_currentEntry);
+    const bool hasTotp = m_currentEntry->hasTotp();
+    m_ui->entryTotpButton->setVisible(hasTotp);
+    m_ui->entryTotpWidget->hide();
+    m_ui->entryTotpButton->setChecked(false);
+
+    if (hasTotp) {
+        m_step = m_currentEntry->totpStep();
+        updateTotp();
+        m_totpTimer->start(m_step * 10);
+    } else {
+        m_totpTimer->stop();
+        m_ui->entryTotpLabel->clear();
+    }
+}
 
-    m_ui->usernameLabel->setText(m_currentEntry->resolveMultiplePlaceholders(m_currentEntry->username()));
+void DetailsWidget::updateEntryGeneralTab()
+{
+    Q_ASSERT(m_currentEntry);
+    m_ui->entryUsernameLabel->setText(m_currentEntry->resolveMultiplePlaceholders(m_currentEntry->username()));
 
     if (!config()->get("security/hidepassworddetails").toBool()) {
-        m_ui->passwordLabel->setText(
-            shortPassword(m_currentEntry->resolveMultiplePlaceholders(m_currentEntry->password())));
-        m_ui->passwordLabel->setToolTip(m_currentEntry->resolveMultiplePlaceholders(m_currentEntry->password()));
+        const QString password = m_currentEntry->resolveMultiplePlaceholders(m_currentEntry->password());
+        m_ui->entryPasswordLabel->setDisplayText(password);
+        m_ui->entryPasswordLabel->setToolTip(password);
     } else {
-        m_ui->passwordLabel->setText("****");
+        m_ui->entryPasswordLabel->setDisplayText("****");
+        m_ui->entryPasswordLabel->setToolTip(QString());
```

Comment:
Use `""`.

Reply:
What are the reasons for using an empty string but not a null string?
[Qt docs -> distinction between null and empty strings](http://doc.qt.io/qt-5/qstring.html#distinction-between-null-and-empty-strings)

Reply:
You can also use {} instead, which will be a null string.

Reply:
> You can also use {} instead, which will be a null string.

done